tests/Tokenizer/TokensTest.php

```diff
@@ -1045,20 +1045,46 @@ public function provideEnsureWhitespaceAtIndexCases()
         );
     }
 
-    public function testRemoveLeadingWhitespace()
+    /**
+     * @dataProvider provideRemoveLeadingWhitespaceCases
+     *
+     * @param int         $index
+     * @param null|string $whitespaces
+     * @param string      $expected
+     * @param string      $input
+     */
+    public function testRemoveLeadingWhitespace($index, $whitespaces, $expected, $input = null)
     {
         Tokens::clearCache();
-        $tokens = Tokens::fromCode('<?php
-// foo
-     $bar;');
 
-        $tokens->removeLeadingWhitespace($tokens->count() - 2);
+        $tokens = Tokens::fromCode(null === $input ? $expected : $input);
+        $tokens->removeLeadingWhitespace($index, $whitespaces);
 
-        $this->assertSame(
-            '<?php
-// foo
-$bar;',
-            $tokens->generateCode());
+        $this->assertSame($expected, $tokens->generateCode());
+    }
+
+    public function provideRemoveLeadingWhitespaceCases()
+    {
+        return array(
+            array(
+                7,
+                " \t",
+                "<?php echo 1;//\n//",
+                "<?php echo 1;//\n       //",
```

Comment:
This is the test that fails that I, as end user of the method, at first found confusing, but in the end makes sense.
In order words; I would update this test to pass and not touch the method.
I think it is up to the caller of the method (for example fixers) to check if it is save to remove the leading space. WDYT?

Reply:
Well, the original failing test was to show that piece of code - `$bar;` - when is directly after one-line comment and `removeLeadingWhitespace` is called for it, will be commented out.

In the meaning of PHP token parser current behaviour is expected, but as we improve the parser to be more reasonable I believe this is the case to do it - I cannot imagine a case when having the current behaviour would be expected.

Reply:
The new test shows the same case and issue as yours, I just put into a provider construct.

So; lets think a bit more about what to expect; lets say for:

(I used `|` to hint a line break and stop GH UI to trim the line)

```
echo 1;  |
          |
 |
     |
   echo 2;
```

If we ask to remove leading spacing before `echo 2;` should it also remove the line breaks because it can be done?
So:

```
echo 1;echo 2;
```

Or will the new version of the method only remove spacing till the first previous line break?
So:

```
echo 1;  |
          |
 |
     |
echo 2;
```

Or remove all spacing before it but always leave it on its own line:

```
echo 1;  |
echo 2;
```

or:
```
echo 1;|
echo 2;
```

When we figured out what we want, we can update :)

Reply:
I'd say `echo 1;echo 2;`.

But it's different when it's with the comment (one-lines of course).

Having:
```php
// hello
    echo 1;
```

The new line after word "hello" is actually part of the comment - let's say if we have a fixer to add `trigger_error` after comment containing `@todo` it would have to move over the new line character and place new tokens "in the middle" of the `T_WHITESPACE` token (if whitespaces were line in the example above). So similarly we should treat mentioned new line when removing leading whitespaces.

Reply:
As creator of the remove(leading|trailing)whitespaces methods...
@SpacePossum , in your example, output shall be `echo 1;echo 2;`.

----

Issue of Kuba is that `//` and `#` aren't contain trailing newline char.
See `WhitespacyCommentTransformer`.

Actually, I would consider this as bug, and fix it by ensuring that this case would work as well:
```php
            array(
                7,
                null,
                "<?php echo 1;//\necho 2;",
                "<?php echo 1;//\n       echo 2;",
            ),
```
(comments are special, removing \n here would change meaning of code)

Reply:
alright lets go with that :+1:

Reply:
please add tests for:
```
echo 1;  |
          |
 |
     |
echo 2;
```
and
```
echo 1;  // 1|
          |
 |
     |
echo 2;
```